Update README.md (#1147)
sparkletown/sparkle: README.md

### Frontend

First, clone the repo and `cd` into it:

```bash
git clone https://github.com/sparkletown/sparkle
cd sparkle
```

> In the main `sparkle` folder, copy `.env.example` to a new file `.env`, and fill it with the appropriate details.

Reviewer: ```suggestion
In the main `sparkle` folder, copy `.env.example` to a new file `.env.local`, and fill it with the appropriate details.
```

> You can read more about `.env` files here: https://create-react-app.dev/docs/adding-custom-environment-variables#adding-development-environment-variables-in-env

Reviewer: ```suggestion
You can read more about the various `.env` files that you can use at:

- https://create-react-app.dev/docs/adding-custom-environment-variables#adding-development-environment-variables-in-env
```

```bash
npm run build
```

### Firebase functions

**Note:** Before you run the following steps, you will need to ensure you have access to the Firebase project you want to use. This access can be set up through the Firebase web UI.

In a new terminal, from the directory you cloned the code to, enter the following commands:

```bash
# While not necessary (as we already include it in our devDependencies), you can install the firebase-tools globally if desired
# npm install -g firebase-tools@latest

# Move into the firebase functions directory
cd functions

# Install the function dependencies
npm install

# Login to firebase with the account that has access to this project
firebase login

# Switch to the 'staging' project in our local configuration (or whichever environment you are developing against)
firebase use staging

# Copy the runtime config locally
npm run firebase functions:config:get > .runtimeconfig.json
```

Now you're ready to launch the backend function emulator! ✨

```bash
npm run firebase:emulate-functions

# Or if you don't want to use our helper scripts, you can do this directly:
# firebase emulators:start --only functions
```

> **Note**: You might need to emulate the firebase functions locally before the server can properly start. If you have issues using/editing the actual staging functions, try that.

Reviewer: ```suggestion

**Note**: You might need to emulate the firebase functions locally before the server can properly start. If you have issues using/editing the actual staging functions, try that.
```

> Clone the repo and cd into it

Reviewer: ```suggestion
First, clone the repo and `cd` into it:
```

```bash
git clone https://github.com/sparkletown/sparkle
cd sparkle
```

> <!-- Now, obtain the `.env` file for the environemnt (eg. staging) and save it (eg. `.env.staging`) -->

Reviewer: No need to comment out these bits, can just delete them.

(only commenting on the first, but this applies to all of the same pattern of issue)

> Now, obtain the `.env` file for the environment (ask someone on the engineering team) and save it.

Reviewer: Since this repo is also open-source, we want to make sure that the instructions are also useful to people outside of Sparkle as well.

Keeping the note about saving it as `.env.development.local` (the previous suggestion of `.env.staging` is wrong) is a good idea.

Potentially we could also link to https://create-react-app.dev/docs/adding-custom-environment-variables#adding-development-environment-variables-in-env to help explain what `.env` files can be used/why/etc.

An empty `.env.example` file exists at https://github.com/sparkletown/sparkle/blob/staging/.env.example that shows the format it needs to be in. So perhaps saying something like about copying `.env.example` to `.env.development.local`, and then filling it in with the appropriate details. For Sparkle, they can get that off other engineers. For random public, the main required variables relate to the deployed firebase project, so they can get them from there (the what/how of deploying this for the first time is probably out of scope (OOS) of this section of the README, so don't worry going into detailed specifics there)

Reviewer: Since we will pretty much always use the same environment for local dev, we probably don't need to mention "for the environment" here either.

<!-- ```
# copy in the .env files
# symlink the staging .env file
ln -s .env.staging.local .env
``` -->

> Install the npm modules

Reviewer: ```suggestion
Install the platform dependencies with `npm`:

(**Note:** `npm` v7+ is not supported, it will cause issues with our `package-lock.json`, and you may end up with the wrong dependency versions)
```

```bash

> npm i

Reviewer: They are equivalent, but this is more 'obvious' what it's doing, so will help new people more

```suggestion
npm install
```

```

Now you're ready to start the server! ✨

```bash

> npm test

Reviewer: We don't need `npm test` here. Technically we don't really have many/any tests at the moment anyways. But it might still be worth mentioning this command somewhere later on, perhaps near the "you can also test builds" section.

```suggestion
```

> npm start

Reviewer: Technically I think that we need to be emulating the firebase functions locally (as below) before the server can properly start/the code work properly. Not 100% about this. It may only be required when we don't want to use the _actual_ staging functions (eg. when developing changes to the functions)

Not sure of the best way to fit that knowledge into this README (or if it's even the right place to). Let's see what you come up with :)

Author: So would someone need to do the firebase steps before the frontend ones?

Author: I did the steps in the order that they came in originally and everything seems to work fine (I think)

Reviewer: > So would someone need to do the firebase steps before the frontend ones?

That was sort of what I was pondering/wondering when I asked my above question.

> I did the steps in the order that they came in originally and everything seems to work fine (I think)

Cool, if it worked that way, then probably should be fine as is :)

```

> You won't need to in dev, but you can also test builds:

Reviewer: ```suggestion
While you generally won't need to do this while developing locally, you can manually build the platform assets as follows:
```

```bash
npm run build
```

### Firebase functions

> To do this, you must first be added to the firebase environment.
> In a new terminal enter the following commands.

Reviewer: ```suggestion
**Note:** Before you run the following steps, you will need to ensure you have access to the Firebase project you want to use. This access can be set up through the Firebase web UI.

In a new terminal, from the directory you cloned the code to, enter the following commands:
```

```bash

> npm install -g firebase-tools@latest

Reviewer: Technically this doesn't need to be installed globally, as we include it as part of our dev dependencies, so it will be installed/usable in that way regardless.

```suggestion
# While not necessary (as we already include it in our devDependencies), you can install the firebase-tools globally if desired
# npm install -g firebase-tools@latest

```

> cd functions
> npm install
> firebase login
> firebase use staging
> firebase functions:config:get

Reviewer: ```suggestion
# Move into the firebase functions directory
cd functions

# Install the function dependencies
npm install

# Login to firebase with the account that has access to this project
firebase login

# Switch to the 'staging' project in our local configuration (or whichever environment you are developing against)
firebase use staging

# Copy the runtime config locally
npm run firebase functions:config:get > .runtimeconfig.json
```

```

> Create a new file `functions/.runtimeconfig.json` and paste the output of this command into it.
> 

Reviewer: Merged into above suggestion.

```suggestion
```

> Now you can launch the server! ✨

Reviewer: ```suggestion
Now you're ready to launch the backend function emulator! ✨
```

<!-- Copy the output of this command and paste it in `functions/.runtimeconfig.json`. Then, launch the server with: -->

```bash

> firebase emulators:start --only functions

Reviewer: ```suggestion
npm run firebase:emulate-functions

# Or if you don't want to use our helper scripts, you can do this directly:
# firebase emulators:start --only functions
```
